include/validator/validator_component.h

```diff
@@ -64,18 +65,67 @@ inline std::string toString(IndexKind IK) {
 
 class Context {
 public:
-  void addComponent(const AST::Component::Component &C) noexcept {
-    CompList.emplace_back(C);
+  struct ComponentContext {
+    const AST::Component::Component *Component;
+    std::vector<AST::Module> CoreModules;
+    std::vector<AST::Component::Component> ChildComponents;
```

Comment:
What if you claim a field `optional<ComponentContext> parent`? Use C++ itself to help you maintain the stack shall be a lot easier.

Also, I suggest you extract a header `include/validator/context.h` out, so internal code will not affect below usage code.

Reply:
In fact, this will let you able to use RAII to replace `enter/exit` component pair.